### ISSUE TITLE
Some countries get stuck in in various merge_ analyzers (block on mapillary content)

### ISSUE THREAD

**Issue by stephankn** [osm-fr/osmose-backend#1191]
Since a few days, analyzers for different countries seem to "get stuck" in the middle. Common part is that they all seem to process input of some sort from mapillary.

I have the impression that a timeout/error handling is missing in handling this external content. Simply blocking forever is not the right solution.

```
2021-02-22 03:12:25 malaysia : merge_traffic_signs
2021-02-22 03:12:26   run osmosis all analyser SubAnalyser_Merge_Traffic_Signs_1
2021-02-22 03:12:26   Batch 1/11: regulatory--maximum-speed-limit-5--g1,regulatory--maximum-speed-limit-10--g1,regulatory--maximum-speed-limit-15--g1,regulatory--maximum-speed-limit-20--g1,regulatory--maximum-speed-limit-30--g1,regulatory--maximum-speed-limit-35--g1,regulatory--maximum-speed-limit-40--g1,regulatory--maximum-speed-limit-45--g1,regulatory--maximum-speed-limit-50--g1,regulatory--maximum-speed-limit-60--g1
2021-02-22 03:12:26   https://a.mapillary.com/v3/map_features?bbox=98.9372144,0.8538205,119.5833462,7.6667327&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&layers=trafficsigns&per_page=1000&start_time=2019-02-23&values=regulatory--maximum-speed-limit-5--g1,regulatory--maximum-speed-limit-10--g1,regulatory--maximum-speed-limit-15--g1,regulatory--maximum-speed-limit-20--g1,regulatory--maximum-speed-limit-30--g1,regulatory--maximum-speed-limit-35--g1,regulatory--maximum-speed-limit-40--g1,regulatory--maximum-speed-limit-45--g1,regulatory--maximum-speed-limit-50--g1,regulatory--maximum-speed-limit-60--g1&sort_by=key
2021-02-22 03:12:26   Page 1
```

```
2021-02-22 05:41:54 taiwan : merge_street_objects
2021-02-22 05:41:54   run osmosis all analyser SubAnalyser_Merge_Street_Objects_2
2021-02-22 05:41:55   Batch 1/1: object--bike-rack,object--cctv-camera,object--fire-hydrant,object--traffic-light--cyclists,object--traffic-light--general-horizontal,object--traffic-light--general-single,object--traffic-light--general-upright,object--traffic-light--other,object--traffic-light--pedestrians,object--phone-booth
2021-02-22 05:41:55   https://a.mapillary.com/v3/map_features?bbox=119.314339,21.8968599,122.1091671,25.6334135&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&layers=points&per_page=1000&start_time=2019-02-23&values=object--bike-rack,object--cctv-camera,object--fire-hydrant,object--traffic-light--cyclists,object--traffic-light--general-horizontal,object--traffic-light--general-single,object--traffic-light--general-upright,object--traffic-light--other,object--traffic-light--pedestrians,object--phone-booth&sort_by=key
2021-02-22 05:41:55   Page 1
2021-02-22 05:42:04   1000 features fetched
2021-02-22 05:42:04   90 keeped
2021-02-22 05:42:04   Page 2
2021-02-22 05:42:11   1000 features fetched
2021-02-22 05:42:11   93 keeped
2021-02-22 05:42:11   Page 3
2021-02-22 05:42:15   1000 features fetched
2021-02-22 05:42:15   88 keeped
2021-02-22 05:42:15   Page 4
2021-02-22 05:42:17   1000 features fetched
2021-02-22 05:42:17   105 keeped
2021-02-22 05:42:17   Page 5
2021-02-22 05:42:21   1000 features fetched
2021-02-22 05:42:21   94 keeped
2021-02-22 05:42:21   Page 6
2021-02-22 05:42:24   1000 features fetched
2021-02-22 05:42:24   114 keeped
2021-02-22 05:42:24   Page 7
```

**Comment by frodrigo** [comment]
The current status of the other servers is good. No issues with Mapillary, nor analysing delay.
It look like the mapillary code use the standard Osmose framework for download. We have done work on it to be false-safe.

When it stops, it never hang up ?
Maybe son kind of limitation from Mapillary on your ip ?

**Comment by stephankn** [comment]
I guess this one blocks:
https://github.com/osm-fr/osmose-backend/blob/21b6c03f050ae812ce220506f7ca4272763e674a/analysers/Analyser_Merge_Mapillary.py#L88

Do we have a similar problem like this?
https://stackoverflow.com/questions/47397919/python-requests-with-httpadapter-is-halting-for-hours

Do we need

    adapter.max_retries.respect_retry_after_header = False

here? 
https://github.com/osm-fr/osmose-backend/blob/21b6c03f050ae812ce220506f7ca4272763e674a/modules/downloader.py#L50

**Comment by stephankn** [comment]
My IP is a static IP in a Hetzner data-center. As osmose is not logging any response headers it is difficult to tell what specifically happened here. I could imagine that it is some sort of rate-limiting from mapillary.

**Comment by stephankn** [comment]
after having killed analysis for malaysia and taiwan, it continued for a while. Now indonesia_riau was blocked for 6 hours in the download process.

**Comment by frodrigo** [comment]
Can you try?
```python
adapter = HTTPAdapter(max_retries=retry, respect_retry_after_header=False)
```
here https://github.com/osm-fr/osmose-backend/blob/master/modules/downloader.py#L50

**Comment by stephankn** [comment]
```
Traceback (most recent call last):
  File "./osmose_run.py", line 386, in run
    return execc(conf, logger, analysers, options, osmosis_manager)
  File "./osmose_run.py", line 166, in execc
    min_file_size=8*1024)
  File "/opt/osmose-backend/modules/download.py", line 56, in dl
    answer = downloader.get(url, headers=headers, session=downloader.requests_retry_session(status_forcelist=downloader.DEFAULT_RETRY_ON + (404, )))
  File "/opt/osmose-backend/modules/downloader.py", line 50, in requests_retry_session
    adapter = HTTPAdapter(max_retries=retry, respect_retry_after_header=False)
TypeError: __init__() got an unexpected keyword argument 'respect_retry_after_header'
Traceback (most recent call last):
  File "./osmose_run.py", line 564, in <module>
    sys.exit(main(options))
  File "./osmose_run.py", line 484, in main
    err_code |= run(country_conf, logger, analysers, options)
  File "./osmose_run.py", line 386, in run
    return execc(conf, logger, analysers, options, osmosis_manager)
  File "./osmose_run.py", line 166, in execc
    min_file_size=8*1024)
  File "/opt/osmose-backend/modules/download.py", line 56, in dl
    answer = downloader.get(url, headers=headers, session=downloader.requests_retry_session(status_forcelist=downloader.DEFAULT_RETRY_ON + (404, )))
  File "/opt/osmose-backend/modules/downloader.py", line 50, in requests_retry_session
    adapter = HTTPAdapter(max_retries=retry, respect_retry_after_header=False)
TypeError: __init__() got an unexpected keyword argument 'respect_retry_after_header'

```

**Comment by stephankn** [comment]
adding the above mentioned line at least builds. I have one analyzer running. Will know tomorrow after the full batch was triggered.

**Comment by stephankn** [comment]
It still blocks in the middle since many hours.

I added some extra code to enable debug output:

```
diff --git a/modules/downloader.py b/modules/downloader.py
index e336d77e..3326a70d 100644
--- a/modules/downloader.py
+++ b/modules/downloader.py
@@ -23,6 +23,8 @@ import hashlib
 import os
 import time
 import requests
+import logging
+import http.client as http_client
 from requests.adapters import HTTPAdapter
 from requests.packages.urllib3.util.retry import Retry
 from datetime import datetime
@@ -39,6 +41,12 @@ DEFAULT_RETRY_ON = (500, 502, 503, 504)


 def requests_retry_session(retries=3, backoff_factor=1, status_forcelist=DEFAULT_RETRY_ON):
+    http_client.HTTPConnection.debuglevel = 1
+    logging.basicConfig()
+    logging.getLogger().setLevel(logging.DEBUG)
+    requests_log = logging.getLogger("requests.packages.urllib3")
+    requests_log.setLevel(logging.DEBUG)
+    requests_log.propagate = True
     session = requests.Session()
     retry = Retry(
         total=retries,
@@ -48,6 +56,7 @@ def requests_retry_session(retries=3, backoff_factor=1, status_forcelist=DEFAULT
         status_forcelist=status_forcelist,
     )
     adapter = HTTPAdapter(max_retries=retry)
+    adapter.max_retries.respect_retry_after_header = False
     session.mount('http://', adapter)
     session.mount('https://', adapter)
     session.headers['User-Agent'] = 'python-requests - https://osmose.openstreetmap.fr/'
```

these are the last log lines before it blocked forever:

```
2021-02-23 02:58:27   Page 11
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): a.mapillary.com:443
DEBUG:urllib3.connectionpool:https://a.mapillary.com:443 "GET /v3/map_features?layers=points&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&bbox=97.3438071%2C5.6128509%2C105.6368119%2C20.4648336&_next_page_token=eyJhZnRlciI6WyJocTF0dXg4bTAzdTF0eXoxY3JpeDNtb3c4NyJdfQ%3D%3D&sort_by=key&start_time=2019-02-24&per_page=1000&values=object--bike-rack%2Cobject--cctv-camera%2Cobject--fire-hydrant%2Cobject--traffic-light--cyclists%2Cobject--traffic-light--general-horizontal%2Cobject--traffic-light--general-single%2Cobject--traffic-light--general-upright%2Cobject--traffic-light--other%2Cobject--traffic-light--pedestrians%2Cobject--phone-booth HTTP/1.1" 200 None
send: b'GET /v3/map_features?layers=points&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&bbox=97.3438071%2C5.6128509%2C105.6368119%2C20.4648336&_next_page_token=eyJhZnRlciI6WyJocTF0dXg4bTAzdTF0eXoxY3JpeDNtb3c4NyJdfQ%3D%3D&sort_by=key&start_time=2019-02-24&per_page=1000&values=object--bike-rack%2Cobject--cctv-camera%2Cobject--fire-hydrant%2Cobject--traffic-light--cyclists%2Cobject--traffic-light--general-horizontal%2Cobject--traffic-light--general-single%2Cobject--traffic-light--general-upright%2Cobject--traffic-light--other%2Cobject--traffic-light--pedestrians%2Cobject--phone-booth HTTP/1.1\r\nHost: a.mapillary.com\r\nUser-Agent: python-requests - https://osmose.openstreetmap.fr/\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 200 OK\r\n'
header: Date: Tue, 23 Feb 2021 02:58:28 GMT
header: Content-Type: application/json; charset=utf-8
header: Transfer-Encoding: chunked
header: Connection: keep-alive
header: access-control-allow-origin: *
header: access-control-expose-headers: Link
header: link: <https://a.mapillary.com/v3/map_features?layers=points&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&values=object--bike-rack%2Cobject--cctv-camera%2Cobject--fire-hydrant%2Cobject--traffic-light--cyclists%2Cobject--traffic-light--general-horizontal%2Cobject--traffic-light--general-single%2Cobject--traffic-light--general-upright%2Cobject--traffic-light--other%2Cobject--traffic-light--pedestrians%2Cobject--phone-booth&sort_by=key&start_time=2019-02-24&per_page=1000&bbox=97.3438071%2C5.6128509%2C105.6368119%2C20.4648336>; rel="first", <https://a.mapillary.com/v3/map_features?layers=points&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&values=object--bike-rack%2Cobject--cctv-camera%2Cobject--fire-hydrant%2Cobject--traffic-light--cyclists%2Cobject--traffic-light--general-horizontal%2Cobject--traffic-light--general-single%2Cobject--traffic-light--general-upright%2Cobject--traffic-light--other%2Cobject--traffic-light--pedestrians%2Cobject--phone-booth&_next_page_token=eyJhZnRlciI6WyJqajcyeWNlZmJsaGNoYjFudGVlb3ZldWlrOCJdfQ%3D%3D&sort_by=key&start_time=2019-02-24&per_page=1000&bbox=97.3438071%2C5.6128509%2C105.6368119%2C20.4648336>; rel="next"
header: etag: W/"1df225-/sashmxUh8/YltLZUA29Wf9XaQY"
header: vary: Accept-Encoding
header: content-encoding: gzip
header: strict-transport-security: max-age=63072000; includeSubdomains;
2021-02-23 02:58:28   1000 features fetched
2021-02-23 02:58:28   990 keeped
2021-02-23 02:58:28   Page 12
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): a.mapillary.com:443
DEBUG:urllib3.connectionpool:https://a.mapillary.com:443 "GET /v3/map_features?layers=points&client_id=MEpmMTFQclBTUWlacjV6RTUxWWMtZzo5OTc2NjY2MmRiMDUwYmMw&values=object--bike-rack%2Cobject--cctv-camera%2Cobject--fire-hydrant%2Cobject--traffic-light--cyclists%2Cobject--traffic-light--general-horizontal%2Cobject--traffic-light--general-single%2Cobject--traffic-light--general-upright%2Cobject--traffic-light--other%2Cobject--traffic-light--pedestrians%2Cobject--phone-booth&_next_page_token=eyJhZnRlciI6WyJqajcyeWNlZmJsaGNoYjFudGVlb3ZldWlrOCJdfQ%3D%3D&sort_by=key&start_time=2019-02-24&per_page=1000&bbox=97.3438071%2C5.6128509%2C105.6368119%2C20.4648336 HTTP/1.1" 200 None
```

Any idea, why not even the "send" output is printed? Does it mean it blocks inside urllib?
Was some version inside docker updated? I wonder why it works most of the time...

How could I further narrow down the issue?

**Comment by stephankn** [comment]
Requests.session is used with `stream=True` in `downloader.py`. 

Documentation says:

> _https://requests.readthedocs.io/en/latest/user/advanced/#body-content-workflow_
> If you set stream to True when making a request, Requests cannot release the connection back to the pool unless you consume all the data or call Response.close. This can lead to inefficiency with connections. If you find yourself partially reading request bodies (or not reading them at all) while using stream=True, you should make the request within a with statement to ensure it’s always closed:

Could it be that the connection pool is exhausted at some conditions? Why do we need `stream=True` within the `Analyser_Merge_Mapillary.py`?

**Comment by frodrigo** [comment]
> Could it be that the connection pool is exhausted at some conditions?

No usage of the pool, only one request at once.

> Why do we need `stream=True` within the `Analyser_Merge_Mapillary.py`?

Not really need of stream for Mapillary, it rather for large dataset, and not really sure we need it.

**Comment by stephankn** [comment]
I am really stuck in trying to understand why it blocks.
With the debugging enabled, the last output I see is a debug message from urllib3. I believe the last line is logged here:
https://github.com/urllib3/urllib3/blob/1.26.3/src/urllib3/connectionpool.py#L452

urllib is version 1.26.3.

`httplib_response.status` is 200, which seems to indicate a valid connection to the server and proper response coming.
`httplib_response.length? is `None`, which is a bit strange, but seems to be normal compared to the previous responses which worked fine.

I think the method got called from here:
https://github.com/urllib3/urllib3/blob/1.26.3/src/urllib3/connectionpool.py#L699

And I think this passes back the result to here:
https://github.com/psf/requests/blob/v2.25.1/requests/adapters.py#L439

I did not find the magic which would add the logging of "send:" which is visible in the good case. I would like to see what is in the code path between these two log outputs in the code-path and what in this path could block. Trying to understand the code without speaking Python is quite hard.

**Comment by frodrigo** [comment]
Try to remove options and the HTTPAdapter from 
https://github.com/osm-fr/osmose-backend/blob/master/modules/downloader.py#L43-L52
and
https://github.com/osm-fr/osmose-backend/blob/master/modules/downloader.py#L60

**Comment by stephankn** [comment]
The first night it worked without blocking. Will check after another night whether reverting the changes brings the problem back.

**Comment by frodrigo** [comment]
which change make it works ?

**Comment by stephankn** [comment]
Second night worked as well. So I reverted the changes (left in the additional debug output) and I manually started a full cycle which worked. Last night worked as well. So I am uncertain how to further proceed.

It is fine, that it no longer blocks. But nothing has changed on osmose side. It suddenly appeared and gone. While the problem existed it was happening quite frequently. I frequently re-created the docker image. I verified the urllib and requests version. But image is fetching probably other components as well which might have impacted it.

The log output showed a return status 200. So I suspect that the problem was happening inside local software and not on remote end. At least the return status arrived, so I expect that other data was coming in as well. Maybe not?

Still this would leave the question why no timeout mechanism killed the connection after a few hours.

**Comment by stephankn** [comment]
I don't think that debugging output makes a change. I also have still in a change which sets stream to False. But before it blocked with this setting as well

```
diff --git a/analysers/Analyser_Merge_Mapillary.py b/analysers/Analyser_Merge_Mapillary.py
index 1c5f7067..4ab4b732 100644
--- a/analysers/Analyser_Merge_Mapillary.py
+++ b/analysers/Analyser_Merge_Mapillary.py
@@ -24,6 +24,7 @@ import json
 from .Analyser_Merge import Source

 import csv
+import pprint
 from modules.PointInPolygon import PointInPolygon
 from modules import SourceVersion
 from modules import downloader
@@ -85,7 +86,8 @@ class Source_Mapillary(Source):
                     while(url):
                         page = page + 1
                         self.logger.log("Page {0}".format(page))
-                        r = downloader.get(url)
+                        r = downloader.get(url, stream=False)
+                        pprint.pprint(r.headers)
                         url = r.links['next']['url'] if 'next' in r.links else None

                         features = r.json()['features']
diff --git a/modules/downloader.py b/modules/downloader.py
index e336d77e..cbf514fa 100644
--- a/modules/downloader.py
+++ b/modules/downloader.py
@@ -23,6 +23,8 @@ import hashlib
 import os
 import time
 import requests
+import logging
+import http.client as http_client
 from requests.adapters import HTTPAdapter
 from requests.packages.urllib3.util.retry import Retry
 from datetime import datetime
@@ -39,6 +41,12 @@ DEFAULT_RETRY_ON = (500, 502, 503, 504)


 def requests_retry_session(retries=3, backoff_factor=1, status_forcelist=DEFAULT_RETRY_ON):
+    http_client.HTTPConnection.debuglevel = 1
+    logging.basicConfig()
+    logging.getLogger().setLevel(logging.DEBUG)
+    requests_log = logging.getLogger("requests.packages.urllib3")
+    requests_log.setLevel(logging.DEBUG)
+    requests_log.propagate = True
     session = requests.Session()
     retry = Retry(
         total=retries,
@@ -48,16 +56,18 @@ def requests_retry_session(retries=3, backoff_factor=1, status_forcelist=DEFAULT
         status_forcelist=status_forcelist,
     )
     adapter = HTTPAdapter(max_retries=retry)
+    adapter.max_retries.respect_retry_after_header = False
     session.mount('http://', adapter)
     session.mount('https://', adapter)
     session.headers['User-Agent'] = 'python-requests - https://osmose.openstreetmap.fr/'
     return session


-def get(url, headers={}, session=None):
+def get(url, headers={}, session=None, stream=True):
     if not session:
         session = requests_retry_session()
-    return session.get(url, headers=headers, stream=True)
+#    return session.get(url)
+    return session.get(url, headers=headers, stream=stream)


 def http_get(url, tmp_file, date_string=None, get=get):

```

**Comment by frodrigo** [comment]
It is now ok for you on this point, can I close this iusse?
Meanwhile, your served is not sending results from 4 days.

**Comment by stephankn** [comment]
Thanks for pointing to the missing updates. While there are no hanging backends listed with `docker ps`, I noticed a lot of docker start commands stuck.

on the 27th, I had this update:

    docker-ce-cli:amd64 (5:20.10.3~3-0~ubuntu-focal

all starts after that seem to be stuck, I have now 91 tasks stuck in this early stage:

    root       24899  0.0  0.0   9608  4540 ?        S    Feb28   0:00 sudo docker-compose -f docker-compose.yml -f ../../osmose-local.yml --project-name osmose_thailand --compatibility run -T --rm backend nice ./osmose_run.py --country=thailand

I looked up docker history and this could relate to this regression: https://github.com/docker/cli/commit/f33a69f6eef118195e741d158cfb8ee7a7391ddb

Yesterday docker released v20.10.5 of docker-cli which fixes this regression. I had to manually clean up stuck processes and networks (bridge is limited to 30 networks).

Right now re-running the analyzers. I expect to have more feedback later today. Also updates should happen to appear./

**Comment by stephankn** [comment]
I have no changes. Problem suddenly came and vanished. I can not point to any change done which was supposed to change it. It is running within docker, for which the image was re-created a few times. It could be related to some ubuntu packages within the docker image.

As it no longer happens, I am closing this issue.  I will leave the debug output in the log for a bit longer. If the issue comes back, we have to think on how we can make osmose non-blocking.
